src/lib/Libifl/tcp_dis_gss.c

```diff
@@ -0,0 +1,1210 @@
+ * @param[out] ebuf - error buffer
+ * @param[in] ebufsz - size of error buffer
+ *
+ * @return	int
+ * @retval	PBS_GSS_OK on success
+ * @retval	!= PBS_GSS_OK on error
+ */
+int
+tcp_gss_client_authenticate(int sock, char *hostname, char *ebuf, int ebufsz)
+{
+	pbs_gss_extra_t *gss_extra = NULL;
+	char *data_in = NULL;
+	int len_in = 0;
+	char *data_out = NULL;
+	int len_out = 0;
+	enum TCP_GSS_MSG_TYPES type;
+	int i;
+	unsigned int data_len;
+	char *data;
+
+	if ((gss_extra = (pbs_gss_extra_t *)tcp_get_extra(sock)) == NULL){
+		gss_extra = pbs_gss_alloc_gss_extra();
+		tcp_set_extra(sock, gss_extra);
+	}
+
+	gss_extra->role = PBS_GSS_CLIENT;
+	gss_extra->req_output_size = DIS_GSS_BUF_SIZE;
+
+	if (gss_extra->hostname == NULL) {
+		gss_extra->hostname = strdup(hostname);
+		if (gss_extra->hostname == NULL) {
+			snprintf(ebuf, ebufsz, "malloc failure");
+			pbs_errno = PBSE_SYSTEM;
+
+			return PBS_GSS_ERR_INTERNAL;
+		}
+	}
+
+	if (tcp_gss_send_auth(sock) != PBS_GSS_OK) {
+		snprintf(ebuf, ebufsz, "authentication failed");
+
+		return PBS_GSS_ERR_INTERNAL;
+	}
+
+	pbs_gss_set_log_handlers(tcp_gss_display_status, tcp_gss_display, NULL);
+
+	do {
+		if (pbs_gss_establish_context(gss_extra, gss_extra->hostname, data_in, len_in, &data_out, &len_out) != PBS_GSS_OK) {
+			snprintf(ebuf, ebufsz, "Failed to establish GSS context");
+			pbs_errno = PBSE_SYSTEM;
+
+			return PBS_GSS_ERR_CONTEXT_ESTABLISH;
+		}
+
+		if (len_in) {
+			free(data_in);
+			len_in = 0;
+		}
+
+		if (len_out) {
+			if (tcp_gss_send_token(sock, TCP_GSS_CTX, data_out, len_out) != PBS_GSS_OK) {
+				snprintf(ebuf, ebufsz, "Failed to assemble GSS context token");
+				pbs_errno = PBSE_SYSTEM;
+				free(data_out);
+
+				return PBS_GSS_ERR_SENDTOKEN;
+			}
+
+			if (DIS_tcp_wflush(sock)) {
+				snprintf(ebuf, ebufsz, "Failed to send GSS context token");
+				pbs_errno = PBSE_SYSTEM;
+				free(data_out);
+
+				return PBS_GSS_ERR_INTERNAL;
+			}
+
+			free(data_out);
+			data_out = NULL;
+			len_out = 0;
+		}
+
+		if (gss_extra->gssctx_established == 0) {
+			/* recieve ctx token */
+
+			i = transport_getc(sock);
+			if (i <= 0) {
+				pbs_errno = PBSE_SYSTEM;
+
+				return PBS_GSS_ERR_RECVTOKEN;
+			}
+			type = (unsigned char)i;
+			if (type != TCP_GSS_CTX) {
+				snprintf(ebuf, ebufsz, "incorrect GSS token type");
+				pbs_errno = PBSE_SYSTEM;
+
+				return PBS_GSS_ERR_RECVTOKEN;
+			}
+
+			data = malloc(sizeof(int));
+			if (data == NULL) {
+				snprintf(ebuf, ebufsz, "malloc failure");
+				pbs_errno = PBSE_SYSTEM;
+
+				return PBS_GSS_ERR_INTERNAL;
+			}
+
+		    	i = transport_gets(sock, data, sizeof(int));
+			if (i <= 0) {
+				pbs_errno = PBSE_SYSTEM;
+
+				return PBS_GSS_ERR_RECVTOKEN;
+			}
+			if (i < sizeof(int)) {
+				snprintf(ebuf, ebufsz, "not enough data to datermine data length");
+				pbs_errno = PBSE_SYSTEM;
+
+				return PBS_GSS_ERR_RECVTOKEN;
+			}
+
+			data_len = ntohl(*((int *) data));
+
+			data = realloc(data, data_len);
```

Comment:
it is safer to use a different pointer for the return value. if realloc() fails, it will return NULL and we will leak memory pointed by data.